doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,879 @@
+ (custom-functions)=
+# Writing Custom Functions
+
+<!-- Note to contributors: if you update one of the examples in this guide, be
+     sure to also update the complete example at the end. -->
+
+This guide will describe how to create custom function classes in SymPy.
+Custom functions use the same mechanisms as the {ref}`functions <functions>`
+that are included with SymPy such as the common {ref}`elementary functions
+<elementary-functions>` like {class}`~.exp()` or {class}`~.sin()`,
+{ref}`special functions <special-functions>` like {class}`~.gamma()` or
+{class}`~.Si()`, and {ref}`combinatorial functions <combinatorial-functions>`
+and {ref}`number theory functions <ntheory-module>` like
+{class}`~.factorial()` or {func}`~.primepi()`.
+
+This guide describes how to define functions that map a subset of
```

Comment:
Suggest describing the goal of this function in words, and/or by giving an example, in addition to in mathematical symbols: someone who wants to write a mathematically simple custom function could get confused by the math already.